File: java/integracion_con_base_de_datos/validar_datos_antes_de_insertar/src/validar_datos_antes_de_insertar/validar_datos_antes_de_insertar.java
package validar_datos_antes_de_insertar;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.util.Scanner;
public class validar_datos_antes_de_insertar {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Connection conn = conexion.conectar();
        if (conn == null) {
            System.out.println("❌ No se pudo conectar a la base de datos");
            sc.close();
            return;
        }
        try {
            System.out.println("=== AGREGAR CONTACTO (CON VALIDACIÓN) ===");
            System.out.print("Nombre: ");
            String nombre = sc.nextLine();
            if (nombre.isEmpty()) {
                System.out.println("❌ El nombre no puede estar vacío");
                sc.close();
                return;
            }
            System.out.print("Teléfono: ");
            String telefono = sc.nextLine();
            if (telefono.isEmpty()) {
                System.out.println("❌ El teléfono no puede estar vacío");
                sc.close();
                return;
            }
            System.out.print("Correo: ");
            String correo = sc.nextLine();
            if (!correo.contains("@")) {
                System.out.println("❌ El correo debe contener @");
                sc.close();
                return;
            }
            String sql = "INSERT INTO contactos (nombre, telefono, correo) VALUES (?, ?, ?)";
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setString(1, nombre);
            ps.setString(2, telefono);
            ps.setString(3, correo);
            int filas = ps.executeUpdate();
            if (filas > 0) {
                System.out.println("✅ Contacto insertado correctamente");
            }
        } catch (Exception e) {
            System.out.println("❌ Error al insertar: " + e.getMessage());
        }
        sc.close();
    }
}

File: visual_basic/manejo_de_errores/validar_datos_antes_de_insertar/validar_datos_antes_de_insertar.vb
Imports System
Imports System.Collections.Generic
Module validar_datos_antes_de_insertar
    Class Contacto
        Public Nombre As String
        Public Telefono As String
        Public Correo As String
    End Class
    Sub Main(args As String())
        Dim agenda As New List(Of Contacto)
        Try
            Dim nuevo As New Contacto()
            Console.Write("Ingrese nombre: ")
            nuevo.Nombre = Console.ReadLine()
            If String.IsNullOrWhiteSpace(nuevo.Nombre) Then
                Throw New Exception("El nombre no puede estar vacío.")
            End If
            Console.Write("Ingrese teléfono: ")
            Dim telefonoInput As String = Console.ReadLine()
            If Not IsNumeric(telefonoInput) Then
                Throw New Exception("El teléfono debe ser numérico.")
            End If
            nuevo.Telefono = telefonoInput
            Console.Write("Ingrese correo: ")
            nuevo.Correo = Console.ReadLine()
            If Not nuevo.Correo.Contains("@") Or Not nuevo.Correo.Contains(".") Then
                Throw New Exception("El correo no es válido.")
            End If
            agenda.Add(nuevo)
            Console.WriteLine("✅ Contacto agregado correctamente.")
        Catch ex As Exception
            Console.WriteLine("❌ Error: " & ex.Message)
        End Try
        Console.ReadKey()
    End Sub
End Module